File: deploy_opdash.py
#!/usr/bin/env python3
import subprocess
import os
import sys
import tarfile
import time

# --- Configuration ---
REMOTE_USER  = "Administrator1"
REMOTE_HOST  = "20.24.58.49"
REMOTE_ROOT  = "/var/www/html/opdash"
SSH_KEY_PATH = os.path.expanduser("~/.ssh/id_rsa")
ARCHIVE_NAME = "opdash-deploy.tar.gz"
ECOSYSTEM_CONFIG = "ecosystem.opdash.config.cjs"
PM2_NAME     = "opdash-backend"

def run_ssh(command: str, timeout=60):
    """Run bundled commands over a single SSH connection with a timeout."""
    ssh_cmd = f'ssh -i "{SSH_KEY_PATH}" -o ConnectTimeout=10 {REMOTE_USER}@{REMOTE_HOST} "{command}"'
    try:
        return subprocess.run(ssh_cmd, shell=True, capture_output=True, text=True, timeout=timeout)
    except subprocess.TimeoutExpired:
        print(f"  [ERROR] SSH command timed out after {timeout}s")
        sys.exit(1)

def main():
    print("\n" + "="*60)
    print("[DEPLOY] OPDASH: DEPLOYMENT (v1.0)")
    print("="*60)
    
    start_time = time.time()

    # 1. Pre-build local assets in client folder
    print("\n[1/5] BUILDING client frontend...")
    env = os.environ.copy()
    env["NODE_OPTIONS"] = "--max-old-space-size=4096"
    try:
        # Run npm run build inside the client subdirectory
        subprocess.run("npm run build", shell=True, check=True, env=env, cwd="client")
    except subprocess.CalledProcessError:
        print("  [ERROR] Client Build failed! Aborting.")
        sys.exit(1)

    # 2. Archive essential files (including client/dist, api, and configs)
    print(f"[2/5] ARCHIVING deployment payload -> {ARCHIVE_NAME}...")
    files_to_include = ["api", "client/dist", "package.json", "package-lock.json", ECOSYSTEM_CONFIG, ".env", "db.json"]
    
    def exclude_node_modules(tarinfo):
        if "node_modules" in tarinfo.name or ".git" in tarinfo.name:
            return None
        return tarinfo

    with tarfile.open(ARCHIVE_NAME, "w:gz") as tar:
        for f in files_to_include:
            if os.path.exists(f):
                tar.add(f, filter=exclude_node_modules)
                print(f"       + {f}")
            else:
                print(f"       [SKIP] not found: {f}")

    print(f"[3/5] UPLOADING archive to {REMOTE_HOST} inside {REMOTE_ROOT}...")
    try:
        # Create directory first to ensure scp works
        run_ssh(f"mkdir -p {REMOTE_ROOT}")
        subprocess.run(f'scp -i "{SSH_KEY_PATH}" {ARCHIVE_NAME} {REMOTE_USER}@{REMOTE_HOST}:{REMOTE_ROOT}/', shell=True, check=True)
    except subprocess.CalledProcessError:
        print("  [ERROR] Upload failed! Check your SSH key and connection.")
        sys.exit(1)

    # 4. Bundled Remote Execution (Self-Healing & Speed)
    print("[4/5] REMOTE extraction, production install, and PM2 reset...")
    remote_script = (
        f"mkdir -p {REMOTE_ROOT}/logs && "
        f"cd {REMOTE_ROOT} && "
        f"tar -xzf {ARCHIVE_NAME} && "
        # Move client/dist/* to root folder for Nginx alias serving after cleaning up old assets
        "rm -rf assets index.html sw.js favicon.png manifest.json && "
        "mv client/dist/* . && "
        "rm -rf client && "
        f"sudo chown -R {REMOTE_USER}:{REMOTE_USER} {REMOTE_ROOT} && "
        # Ensure DATABASE_URL connects directly to Azure Postgres (port 5432) since PgBouncer (6432) has no OpDash route
        "sed -i 's/20.24.58.49:6432/stride-posgre-prod-01.postgres.database.azure.com:5432/g' .env && "
        "echo \"       → Running production npm install...\" && "
        "npm install --omit=dev --legacy-peer-deps --prefer-offline --no-audit --no-fund 2>&1 | tail -n 10 && "
        f"pm2 flush {PM2_NAME} && "
        f"pm2 delete {PM2_NAME} 2>/dev/null || true && "
        f"pm2 start {ECOSYSTEM_CONFIG} && "
        f"rm {ARCHIVE_NAME}"
    )
    
    # Use -t to force a pseudo-terminal, which helps with process cleanup
    ssh_cmd = f'ssh -t -i "{SSH_KEY_PATH}" -o ConnectTimeout=10 {REMOTE_USER}@{REMOTE_HOST} "{remote_script}"'
    try:
        # We don't use capture_output here so the user can see the progress (echoes, etc.)
        subprocess.run(ssh_cmd, shell=True, check=True)
    except subprocess.CalledProcessError as e:
        print(f"  [ERROR] Remote setup failed with exit code {e.returncode}")
        sys.exit(1)
    except KeyboardInterrupt:
        print("\n  [WARN] Deployment interrupted by user.")
        sys.exit(1)

    # 5. Final Health Check
    print("[5/5] VERIFYING remote API health...")
    verify_cmd = "curl -s http://127.0.0.1:3001/health || curl -s http://127.0.0.1:3001/api/health || curl -s http://127.0.0.1:3001/"
    health = run_ssh(verify_cmd)
    if health.stdout.strip() != "":
        print("      SUCCESS: API confirmed online on port 3001!")
    else:
        print(f"      [WARN] Health check warning: Unexpected response. Check PM2 logs.")

    # Local cleanup
    if os.path.exists(ARCHIVE_NAME):
        os.remove(ARCHIVE_NAME)

    duration = time.time() - start_time
    print("\n" + "="*60)
    print(f"[SUCCESS] Deployment Success! (Duration: {duration:.1f}s)")
    print("    URL: https://stride.deped.gov.ph/opdash/")
    print("="*60 + "\n")

if __name__ == "__main__":
    main()

File: deploy_opdash.js
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const REMOTE_USER  = "Administrator1";
const REMOTE_HOST  = "20.24.58.49";
const REMOTE_ROOT  = "/var/www/html/opdash";
const SSH_KEY_PATH = path.join(process.env.USERPROFILE || process.env.HOME || '', '.ssh/id_rsa');
const ARCHIVE_NAME = "opdash-deploy.tar.gz";
const ECOSYSTEM_CONFIG = "ecosystem.opdash.config.cjs";
const PM2_NAME     = "opdash-backend";

function runLocal(cmd, cwd) {
    console.log(`Executing: ${cmd}`);
    try {
        execSync(cmd, { stdio: 'inherit', cwd, env: { ...process.env, NODE_OPTIONS: "--max-old-space-size=4096" } });
    } catch (err) {
        console.error(`Error executing command: ${cmd}`, err);
        process.exit(1);
    }
}

function main() {
    console.log("\n============================================================");
    console.log("[DEPLOY] OPDASH: DEPLOYMENT VIA NODE.JS (v1.0)");
    console.log("============================================================\n");

    // 1. Build frontend
    console.log("[1/5] Building client frontend...");
    runLocal("npm run build", "client");

    // 2. Archive files using system tar
    console.log(`[2/5] Archiving deployment payload -> ${ARCHIVE_NAME}...`);
    const filesToInclude = ["api", "client/dist", "package.json", "package-lock.json", ECOSYSTEM_CONFIG, ".env", "db.json"];
    
    // Remove old archive if exists
    if (fs.existsSync(ARCHIVE_NAME)) {
        fs.unlinkSync(ARCHIVE_NAME);
    }
    
    // Run system tar
    runLocal(`tar -czf ${ARCHIVE_NAME} ${filesToInclude.join(' ')}`);

    // 3. Upload archive
    console.log(`[3/5] Uploading archive to ${REMOTE_HOST} inside ${REMOTE_ROOT}...`);
    const sshKeyOption = fs.existsSync(SSH_KEY_PATH) ? `-i "${SSH_KEY_PATH}"` : '';
    
    // Ensure remote directory exists
    const mkdirCmd = `ssh ${sshKeyOption} -o ConnectTimeout=10 ${REMOTE_USER}@${REMOTE_HOST} "mkdir -p ${REMOTE_ROOT}"`;
    runLocal(mkdirCmd);
    
    // Copy file
    const scpCmd = `scp ${sshKeyOption} ${ARCHIVE_NAME} ${REMOTE_USER}@${REMOTE_HOST}:${REMOTE_ROOT}/`;
    runLocal(scpCmd);

    // 4. Remote execution
    console.log("[4/5] Remote extraction, production install, and PM2 reset...");
    const remoteScript = [
        `mkdir -p ${REMOTE_ROOT}/logs`,
        `cd ${REMOTE_ROOT}`,
        `tar -xzf ${ARCHIVE_NAME}`,
        `rm -rf assets index.html sw.js favicon.png manifest.json`,
        `mv client/dist/* .`,
        `rm -rf client`,
        `sudo chown -R ${REMOTE_USER}:${REMOTE_USER} ${REMOTE_ROOT}`,
        `sed -i 's/20.24.58.49:6432/stride-posgre-prod-01.postgres.database.azure.com:5432/g' .env`,
        `echo "       → Running production npm install..."`,
        `npm install --omit=dev --legacy-peer-deps --prefer-offline --no-audit --no-fund 2>&1 | tail -n 10`,
        `pm2 flush ${PM2_NAME}`,
        `pm2 delete ${PM2_NAME} 2>/dev/null || true`,
        `pm2 start ${ECOSYSTEM_CONFIG}`,
        `rm ${ARCHIVE_NAME}`
    ].join(' && ');

    const sshExecuteCmd = `ssh -t ${sshKeyOption} -o ConnectTimeout=10 ${REMOTE_USER}@${REMOTE_HOST} "${remoteScript}"`;
    runLocal(sshExecuteCmd);

    console.log("\n============================================================");
    console.log("[DEPLOY] SUCCESSFUL!");
    console.log("============================================================\n");
}

main();
